packages/peregrine/lib/talons/CheckoutPage/useCheckoutPage.js

```diff
@@ -0,0 +1,95 @@
+import { useCallback, useState } from 'react';
+import { useMutation } from '@apollo/react-hooks';
+
+import { useAwaitQuery } from '../../hooks/useAwaitQuery';
+import { useAppContext } from '../../context/app';
+import { useUserContext } from '../../context/user';
+import { useCartContext } from '../../context/cart';
+
+export const useCheckoutPage = props => {
+    const { createCartMutation, getCartDetailsQuery } = props;
+
+    const [, { toggleDrawer }] = useAppContext();
+    const [{ isSignedIn }] = useUserContext();
+    const [
+        { isEmpty },
+        { createCart, getCartDetails, removeCart }
+    ] = useCartContext();
+
+    const [fetchCartId] = useMutation(createCartMutation);
+    const fetchCartDetails = useAwaitQuery(getCartDetailsQuery);
+
+    const handleSignIn = useCallback(() => {
+        toggleDrawer('nav');
+    }, [toggleDrawer]);
+
+    /**
+     * TODO. This needs to change to checkout mutations
+     * or other mutations once we start checkout development.
+     *
+     * For now we will be using removeCart and createCart to
+     * simulate a cart clear on Place Order button click.
+     */
+    const cleanUpCart = useCallback(async () => {
+        await removeCart();
+
+        await createCart({
+            fetchCartId
+        });
+
+        await getCartDetails({ fetchCartId, fetchCartDetails });
+    }, [removeCart, createCart, getCartDetails, fetchCartId, fetchCartDetails]);
+
+    /**
+     * Using local state to maintain these booleans. Can be
+     * moved to checkout context in the future if needed.
+     *
+     * These are needed to track progree of checkout steps.
+     */
+    const [shippingInformationDone, updateShippingInformationDone] = useState(
+        false
+    );
+    const [shippingMethodDone, updateShippingMethodDone] = useState(false);
+    const [paymentInformationDone, updatePaymentInformationDone] = useState(
+        false
+    );
+    const [orderPlaced, updateOrderPlaced] = useState(false);
+
+    const setShippingInformationDone = useCallback(
+        () => updateShippingInformationDone(true),
+        [updateShippingInformationDone]
+    );
+    const setShippingMethodDone = useCallback(
+        () => updateShippingMethodDone(true),
+        [updateShippingMethodDone]
+    );
+    const setPaymentInformationDone = useCallback(
+        () => updatePaymentInformationDone(true),
+        [updatePaymentInformationDone]
+    );
+    /**
+     * TODO
+     */
+    const placeOrder = useCallback(async () => {
+        await cleanUpCart();
+        updateOrderPlaced(true);
+    }, [cleanUpCart, updateOrderPlaced]);
+
+    return [
+        {
```

Comment:
While we use the `[state, api]` return for context hooks, no other talons use this API that I'm aware of.

I really do like the idea, but for now maybe we should stick to the norm? What are your thoughts @jimbo? I feel like this makes the talon APIs much more resistant to breakage.

Reply:
Yeah, I was confused of this as well. I personally feel `[state, api]` is a cleaner structure. I might be wrong but react (un-officially) pushes that return structure for react hooks (both built in and custom).

Reply:
> Yeah, I was confused of this as well. I personally feel `[state, api]` is a cleaner structure. I might be wrong but react (un-officially) pushes that return structure for react hooks (both built in and custom).

You're correct, React does favor a `[state, api]` return signature in its own hooks. That also happens to be the reason why we selected that signature for our context hooks, such as `useAppContext`. It's a very intuitive structure.

Talons are a bit different though, so we took them in a different direction. They're more like components, in that they can do pretty much anything; they can hold state, define callbacks, and perform side effects. The only thing they can't do is create elements. In order to not confine them, and to ensure they don't appear to be generic or reusable, we gave them the most basic, consistent API possible:

```es6
propsObjectIn => propsObjectOut
```

It's not that talons are incompatible with a `[state, api]` signature, though; it's that it's too late to change the pattern (for now). In the future, we can revisit the talon APIs **as a whole** and consider whether to apply more constraints to them or reshape the signature. Just now, though, adoption of this new feature will depend on consistency, so let's stick to the existing pattern. Object in, object out, please.